### PR TITLE
Add unified Spatial Mapping Components

### LINKED ISSUE #4: Create unified Spatial Mapping Components
We should have components to handle rendering and creating colliders for spatial mapping.

These components should be usable by default but customizable to what most people need. Additionally, we should have heuristics to handle Tracking Lost and moving far away from a placed observer intelligently instead of just removing the meshes.

### PR DESCRIPTION
**Pull request by jim-wordelman-msft**
This adds two components: SpatialMappingCollider and
SpatialMappingRenderer. These components provide intelligent defaults
and follow best practices for handling Spatial Mapping.

Additionally, these components are customizable and implement heuristics
that better support tracking loss or moving far away from an observed
region.

Additionally

 - Move Wireframe and Occlusion shaders and materials to
   Resources/HoloToolkit for default loading in the components

 - Add section to README.md describing these components and update
   existing SpatialMapping section to suggest using the components by
   default.

This fixes #4

### CODE REVIEW

**Comment by robkelly** [comment]
Reviewed and signed off on components in codeflow prior to pull request.

File: Assets/HoloToolkit/SpatialMappingComponent/SpatialMappingRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    /// <summary>
+    /// The material used to render the mesh if _currentRenderingSetting is RenderingSetting.Occlusion
+    /// </summary>
+    private Material OcclusionMaterial = null;
+
+    protected override void Start()
+    {
+        base.Start();
+
+        OcclusionMaterial = Resources.Load("MsSpatialMappingComponent/Occlusion", typeof(Material)) as Material;
```
Comment (fieldsJacksonG) [inline Assets/HoloToolkit/SpatialMappingComponent/SpatialMappingRenderer.cs]: HoloToolkit/Occlusion

File: Assets/HoloToolkit/SpatialMappingComponent/SpatialMappingRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+        if (RenderingMaterial == null)
+        {
+            RenderingMaterial = Resources.Load("MsSpatialMappingComponent/Wireframe", typeof(Material)) as Material;
```
Comment (fieldsJacksonG) [inline Assets/HoloToolkit/SpatialMappingComponent/SpatialMappingRenderer.cs]: HoloToolkit/Wireframe

File: Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+    /// <summary>
+    /// Handler for when the SurfaceObserver completes RequestMeshAsync
+    /// 
+    /// The base class defines this function to make it easier for subclasses to modify meshes upon completion without needing to handle the actual processing
+    /// </summary>
+    /// <param name="bakedData">The processed SurfaceData</param>
+    /// <param name="outputWritten">Whether or not output was written</param>
+    /// <param name="elapsedBakeTimeSeconds">The time in seconds it took to request and populate the mesh</param>
+    protected virtual void SurfaceObserver_OnDataReady(SurfaceData bakedData, bool outputWritten, float elapsedBakeTimeSeconds)
+    {
```
Comment (fieldsJacksonG) [inline Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs]: Please mark this as a deliberate passthrough

File: Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+            MeshFilter filter = surfaceObject.GetComponent<MeshFilter>();
+            if (filter != null)
+            {
+                if (filter.mesh != null)
+                {
+                    Destroy(filter.mesh);
+                }
+            }
+
+            Destroy(surfaceObject);
```
Comment (NeerajW) [inline Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs]: Destroy vs DestroyImmediate perhaps?

File: README.md [hunk 5]
```diff
@@ -106,8 +106,74 @@ or in your Visual Studio Package.appxmanifest capabilities.
 
 ---
 
+## SpatialMappingComponent
```
Comment (NeerajW) [inline README.md]: Could you also please add a test in Tests folder? That really helps with avoiding regressions when others change code. Also a good way to show usage.

File: Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs (hunk 4 above)
Comment (jim-wordelman-msft) [inline Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs]: I feel Destroy is more appropriate here. We don't have a need to remove these objects this instant (as in before the end of the frame) and Destroy is generally seen as safer

File: Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs (hunk 3 above)
Comment (jim-wordelman-msft) [inline Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs]: I'm sorry, I'm not familiar with that concept. Could you explain more/link to an explanation/post example code?
Thanks

Comment (jim-wordelman-msft) [inline Assets/HoloToolkit/SpatialMappingComponent/Support/SMBaseAbstract.cs]: Resolved offline and integrated into latest push

**Pull request merged**